### ISSUE TITLE
Load data little-by-little

### ISSUE THREAD

**Issue by nicolas-raoul** [alemela/wiki-needs-pictures#10]
The CSV file is currently around 7 MB, it is huge for mobile users. Mobile users are probably the people who want to use wiki-needs-pictures the most.

The quantity of data will continue to grow, and could even reach several hundred megabytes if Wikidata requested images are included.

So, rather than loading 7 MB upfront, how about splitting the data in regions, and only loading the regions that are around me? That will probably require abandoning the clustering gadget, but I believe this is needed.

**Comment by alemela** [comment]
I'm working just on that. I split the big data csv file in countries so at the start you have to choose which region to load or simply click "near me". I've arrived to your same conclusion while importing wikidata (so also that script is finished but for now unusable).

**Comment by nicolas-raoul** [comment]
Rather than splitting by country, I would suggest splitting in "squares".
Using squares makes calculations much easier.

First, decide the size of the squares so that the most crowded square has
about 1MB worth of points. Generating the files is easy using
latitude/longitude of each point.

1. From the client, get the user's location.
2. Judging from the current zoom and window size, download the necessary
squares.
3. When the user moves, download more squares if necessary.

(Another strategy can be to send latitude1,longitude1,latitude2,longitude2
from the client and send a single file containing all data for that zone.
Faster download, but more server-side concatenations.)

For zoom levels lower than X, download a special data file that only
contains the number of points per square, and display these numbers on the
map (that's clustering).

**Comment by alemela** [comment]
I agree that this scenario is the best way to represent huge amount of data. But I've chosen to split by country because all the work is done serverside and the user have just to download a single file at the beginning and the calculations are terrible easy. Morehover during desktop experience maybe you don't want to share your location and the problem of loading all data remains. That's because I can't send only summary numbers to the cluster library, I've to use or write another one if needed.

So that's why I've started with a compromise, but surely I'll maintain this issue open as it is the best way in the long run.

**Comment by nicolas-raoul** [comment]
The Russia file will probably be huge though? Will you sub-split it? And at
the opposite, someone in Luxemburg (or near any Schengen border) will be
inconvenienced.

Anyway, good luck and looking forward to the app being able to at least
handle all the data :-)

**Comment by nemobis** [comment]
Ah, now I understand why I have to select a country at the beginning. So far no problems for me.
